### DISCUSSION TITLE
Why we need /plugins folder for nodejs wiki? Can we just use control panel to install/upgrade plugins?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6096] [General]
One of the drawbacks of nodejs wiki (and editions that based on it, for example, TiddlyGit) is that it is hard to update and install plugins for a non-technical user, because it requires people to 1. find the folder 2. open info file with some text editor and edit the json (that is somehow geek for most people).

I hope install/update plugins can be done just by clicking on a button (like how it can be done in a single HTML wiki). This makes me wonder, why do we need that `/plugins` folder? Can we make the plugin experience exactly the same in the HTML wiki and nodejs wiki? 

As @Jermolene said in https://github.com/Jermolene/TiddlyWiki5/issues/4570#issuecomment-616513642 , this is because we don't want some plugins to be copied to the tiddlers folder and makes upgrade difficult, but there is an upgrade button in the "control panel - plugins" now, so this might not be a problem. And following this parading makes installing and upgrading 3rd party plugins also difficult for normal people.

I think we can improve this workflow by two ways:

### Auto unpack json plugin

1. Allow users to install built-in plugins by clicking "enable" button in the "control panel - plugins"
2. When install plugins as JSON tiddlers, by via drag&drop or click "install" in the "control panel - plugins" (after drag in a 3rd plugin library), we check is this a built-in plugin, if so, we modify the info file automatically.
3. If it is a 3rd plugin, we transform JSON tiddler and move generated files (generated by `--savewikifolder <wikifolderpath> [<filter>]`) into plugins folder, and delete the original JSON plugin file.

Or 

### Just use json plugin

1. delete the /plugins folder, and just ignore the `plugins` section in the `tiddlywiki.info`
2. just let user install json plugin, don't mind the duplication, because plugins are not too big
3. We have "upgrade" button in the plugin panel, so upgrade built-in plugins is not a problem any more

I just don't know will JSON plugin format break some plugin functionality? Like js file in the plugin that requires `fs` (for example, plugins that update wiki when fs changes) or js that uses __dirname . And will this make JS's error message hard to debug?

**Comment by pmario** [comment]
@Are you sure, you want to install untrusted 3rd party plugins with a 1 click install onto your harddisk, that will allow your server to access your system with the same rights as you?

IMO it would be the same as if you post your bank account and your pin here

**Reply by linonetwo** [reply]
I think the community plugin library can solve this, we review and can revoke some misused plugins from it. This is how npm solves this problem, npm package may be harmful too.

**Reply by linonetwo** [reply]
We are using pull request to accept plugins into our Chinese community plugin library now.

And I think our auditing works basically fine during this period...

**Comment by Jermolene** [comment]
Hi @linonetwo some of the complexity here stems from the way that in the client server configuration we actually run two copies of TiddlyWiki: one in the browser, and one under Node.js. The requirement to install plugins via the `/plugins` folder of the wiki only applies if one needs the plugin to work under Node.js. In practice, the vast majority of third party plugins only need to be run in the browser, and so don't need to be installed via the `/plugins` folder. Instead, one can just install such plugins by dragging and dropping the JSON file, or using a plugin library.

**Reply by linonetwo** [reply]
Thanks for the clarification, I think I shouldn't use the `/plugins` folder for most of the cases. It is for advanced users who understand the danger that @pmario points out, and they should open the wiki via the file explorer, and edit the `.info` file manually.

And I will add back the OfficialPluginLibrary back in TiddlyGit, as OfficialPluginLibrary allows a normal user to upgrade and install official plugins by clicking a button. Which I think is important when advertising TiddlyWiki. Manual clicking button to upgrade plugins/ the redundancy in the disk space, are not the problem compared to the difficulty of editing the `.info` file.  ( I can add a note that suggests the user to edit the `.info` file, but still allows just click on button )

And, very looking forward to the community plugin library, if we can move all the official plugins to that CPL, this won't be a problem too.

**Reply by linonetwo** [reply]
Do you know when will the CPL come out? We are going to make one for the Chinese community first, in https://github.com/tiddly-gittly/TiddlyWiki-Chinese-Tutorial , so TiddlyGit can enjoy the easy plugin update before CPL comes out.

**Reply by linonetwo** [reply]
I really enjoy the Chinese CPL these days! 

After push code to the repo, github actions will build plugin library within 1 min, and all my wikis receive the update. And can upgrade with one click.

This plugin library feature and architcture in the TiddlyWiki is really a great design!
